### ISSUE TITLE
Should the groups be object-oriented?

### ISSUE THREAD

**Issue by KOLANICH** [tomerfiliba/plumbum#311]
>If you wish to group certain switches together in the help message, you can specify group = "Group Name", where Group Name is any string.

>COLOR_GROUPS = {"Meta-switches" : colors.bold & colors.yellow}

It can be a bit inconvenient and looks a bit ugly. Should we encapsulate all these in ```Group``` class?

**Comment by henryiii** [comment]
Then you would have to define groups at the top, and then pass that to each switch, etc. instead of group="MyGroup". I don't think this is much better, and makes simple, uncolored groups much harder to use. Feel free to home up with a pseudo example, and see if you can get a nice syntax.

Remember, you need to be able to change the default groups colors, as well...

Current color:

~~~python
class MyApp(cli.Application):
    COLOR_GROUPS = {"Meta-switches": colors.bold & colors.yellow,
                    "Mine": colors.green}
    a = cli.Flag("-a", group="Mine", help = "All")
    ...
~~~

Current no color:

~~~python
class MyApp(cli.Application):
    a = cli.Flag("-a", group="Mine", help = "All")
    ...
~~~

Maybe this would be possible with some sort of `add_group` function?

**Comment by KOLANICH** [comment]
>Then you would have to define groups at the top, and then pass that to each switch, etc. instead of group="MyGroup".

See no problem with it.

>I don't think this is much better, and makes simple, uncolored groups much harder to use.

It should benefit readability

```python
class MyApp(cli.Application):
	class first(cli.Group):
		name="First group"
		description="The first group"
		color=colors.bold & colors.yellow
		a=cli.Flag("-a", help = "a")
	
	class second(cli.Group):
		name="Second group"
		description="The second group"
		b=cli.Flag("-b", help = "b")
	
	c = cli.Flag("-c", help = "c")
	
	def main(self):
		print(self.first.a, self.second.b, self.c)
```

**Comment by henryiii** [comment]
Wow, not what I was thinking. I rather like that. Name, description, and color should be all caps to match cli (indicates that the name is special). Other than that, I like it. Do you think you could prepare a PR? I won't have time for a few weeks, I expect.

**Comment by henryiii** [comment]
Also, doesn't directly show what is done with Meta-switches, etc, but that could be added trivially, I think.

**Comment by henryiii** [comment]
Could the name just come from the class name? That's one feature of using a class. Is there a reason to have a group with spaces in the name?

**Comment by henryiii** [comment]
Also, the help description should be like `cli.Application`, and support using the class doc-string for description too (I added that to your example).

**Comment by henryiii** [comment]
And, since now the name and description are defined in a single place, there's no reason to have a separate `COLOR` property at all, just set the name/description with a color (Might be a reason to have  name property). I would default to using the class name if no `NAME` was present.

I'm liking this :)

**Comment by KOLANICH** [comment]
>Could the name just come from the class name

I think, yes, if ```NAME``` is missing.

>Also, the help description should be like cli.Application, and support using the class doc-string for description too (I added that to your example).

We can drop ```DESCRIPTION``` then.

**Comment by henryiii** [comment]
Like in `cli.Application`, it's important to have `DESCRIPTION` as a fallback since there are times when you can't use the help line, since you can't use dynamically generated content in the doc-string (IIRC). I don't think you can use `textwrap.dedent`, for example. But in normal cases, it's fine.

**Comment by KOLANICH** [comment]
>since you can't use dynamically generated content in the doc-string (IIRC).

Of course I can

```python
class B:
    __doc__=str(1+2)
b=B()
?b
#Type:        B
#Docstring:   3
```

**Comment by henryiii** [comment]
Of course... Might have to look into why that was done that way for `Application`, then. I'm fine with dropping `DESCRIPTION`.

**Comment by KOLANICH** [comment]
Control codes?

**Comment by henryiii** [comment]
An example would help (for your library too :) ).

**Comment by KOLANICH** [comment]
The example is in the library itself (in the bottom of py file).
The way it can be helpful to solve the problem with color codes is that it stores the message in hierarchical way, so the tools like jupyter can easily transform it into representation suitable to show in the needed environment.
